sjsonnet/src/sjsonnet/Val.scala

```diff
@@ -212,6 +212,13 @@ object Val{
         .defaults
         .map{ case (index, default) => (index, Lazy(evalDefault(default, newScope, evaluator)))}
 
+      // Check if expected args match passed args before  
+      if (params.args.length < args.length) {
```

Comment:
Why not make this more general, if the number of arguments `!=`, instead of only for too many arguments? The error would be something like `"Wrong number of arguments, expected ${params.args.length} but ${args.length} found"`

Reply:
Apparently, there are valid use cases in which the number might be different, however, I couldn't catch any case where the number is greater. If we change the comparison to `!=`, the following tests start to fail:

```
X sjsonnet.StdWithKeyFTests.stdSetMemberWithKeyF 185ms
X sjsonnet.StdWithKeyFTests.stdSortWithKeyF 1ms
X sjsonnet.StdWithKeyFTests.stdUniqWithKeyF 1ms
X sjsonnet.StdWithKeyFTests.stdSetWithKeyF 1ms
X sjsonnet.StdWithKeyFTests.stdSetUnionWithKeyF 1ms
X sjsonnet.StdWithKeyFTests.stdSetInterWithKeyF 1ms
X sjsonnet.StdWithKeyFTests.stdSetDiffWithKeyF 1ms
X sjsonnet.FileTests.functions 11ms
X sjsonnet.FileTests.invariant_manifest 2ms
X sjsonnet.FileTests.parsing_edge_cases 2ms
X sjsonnet.FileTests.stdlib 99ms
X sjsonnet.FileTests.tla.simple 1ms
X sjsonnet.PreserveOrderTests.preserveOrderYaml 0ms
+ sjsonnet.PreserveOrderTests.preserveOrderXml 0ms
X sjsonnet.PreserveOrderTests.preserveOrderPreservesSetMembership 0ms
X sjsonnet.ErrorTests.invariant.option 1ms
X sjsonnet.ErrorTests.import_wrong_nr_args 1ms
X sjsonnet.ErrorTests.too_many_arg_with_named_arg 0ms
+ sjsonnet.FormatTests.X 2ms
X sjsonnet.EvaluatorTests.functions 0ms
X sjsonnet.EvaluatorTests.members 1ms
X sjsonnet.EvaluatorTests.topLevel 0ms
X sjsonnet.EvaluatorTests.lazy 1ms
X sjsonnet.EvaluatorTests.unboundParam 1ms
```

The minority seems to be because there is now order for the checks, so tests that were asserting against wrongly bound arguments, but passing too many args to the function call, fail with the wrong error stack trace. But the majority seems to be a valid case where the number of arguments does not match - I can't really tell why. For example:

```
test("functions") {
      eval("(function(x) x)(1)") ==> ujson.Num(1)
      eval("function() 1") ==> ujson.Num(1)
      eval("function(a=1) a") ==> ujson.Num(1)
      eval("(function(x, y = x + 1) y)(x = 10)") ==> ujson.Num(11)
}
```

I'm not sure if this is expected or not

Reply:
Could it be that my suggestion was too naive? Maybe the number of arguments at the call site should be tested against the number of parameters without a default value? Not sure if the implementation follows the specification (with a desugaring step that rewrites all functions to have a default value). Anyway, if it's getting too hairy let's leave it as it was

Reply:
Following up offline, later I can place our conclusions here for the sake of documenting it